### PR TITLE
Challenges

### PR DESCRIPTION
**Pull request by PatSCoder**
Jag har lagt till funktion för att kunna visa dom tre korten.

### CODE REVIEW

**Comment by DanneCarlson** [comment]
Fint.

**Review by DanneCarlson — APPROVED** [review]

File: 02-esc/index.js [hunk 1]
```diff
@@ -1,7 +1,23 @@
 import { getDataFromAPI } from "./getDataFromAPI.js";
+import { showAllChallenges } from "./showAllChallenges.js";
 
-// getDataFromAPI returns an array with all the API data
-// console.log(await getDataFromAPI());
+const challenge_list = document.querySelector('.challenge-list');
```
Comment (richardolsson) [inline 02-esc/index.js:4]: This name (`challenge_list`) uses a casing that is inconsistent with most other names in your code. This is called "snake case" (`lowercase_words_with_underscores`) and the norm in your code (and javascript at large) is "camel case" (`everyWordExceptTheFirstIsCapitalized`). 

I think the name of this variable should have been `challengeList`, which would have made it consistent with the rest of the code.

File: 02-esc/index.js [hunk 2]
```diff
@@ -1,7 +1,23 @@
...
+
+let resFromAPI = await getDataFromAPI();
```
Comment (richardolsson) [inline 02-esc/index.js:6]: I think this should be a `const` variable, instead of a `let`. Constants can't change, so if you're not planning to change the value of a variable (like here, this doesn't change), using `const` instead of `let` means the language will help you discover if you deviate from that plan, by telling you you're not allowed.

If later it turns out you do need to reset the value somewhere, changing to ´let` is simple. For that reason, it's always best to start with `const` unless you know that a variable will need to be changing.

File: 02-esc/index.js [hunk 3]
```diff
@@ -1,7 +1,23 @@
...
+
+function sortRatingArray(resFromAPI){
```
Comment (richardolsson) [inline 02-esc/index.js:8]: The name of this function implies to me that it just sorts the array, but it also shows it. I would have refactored this function to return the sorted array, and then use `showAllChallenges()` with the return value, instead of having the function that implies just sorting also show it.

With my changes (ignoring any changes I would make to names), here's what calling the `sortRatingArray()` and `showAllChallenges()` functions would look like:

```js
const topThreeChallenges = sortRatingArray(resFromAPI);
showAllChallenges(topThreeChallenges, challenge_list);
```

File: 02-esc/index.js [hunk 4]
```diff
@@ -1,7 +1,23 @@
...
+let resFromAPI = await getDataFromAPI();
+
+function sortRatingArray(resFromAPI){ 
+
+  resFromAPI.sort(function(a, b){
+    return b.rating - a.rating
+  });
+    
+  let newArray = [];
+  newArray.push(resFromAPI[0], resFromAPI[1], resFromAPI[2]);
```
Comment (richardolsson) [inline 02-esc/index.js:15]: The following lines of code do the same thing, but are better IMO:

**Option 1**
```js
const newArray = [ resFromAPI[0], resFromAPI[1], resFromAPI[2] ];
```

**Option 2**
```js
const newArray = resFromApi.slice(0, 3);
```

File: 02-esc/index.js (hunk 3 above)
Comment (richardolsson) [inline 02-esc/index.js:8]: The name of the `resFromAPI` parameter indicates to me that maybe you are not yet comfortable with thinking about functions as vehicles for separation of concerns.

This function has nothing to do with the API. It doesn't really care where the list of challenges comes from, as long as it has a certain shape (challenges need to have a `rating` property, and be passable to  `showAllChallenges()`).

Inside this function, the source is irrelevant, and the name of the parameter is only used inside of this function, so why not call it something simpler, e.g. simply `challenges` instead of `resFromAPI`.

This also helps to separate, in the mind of the reader of your code, the `resFromAPI` name in this function from the `resFromAPI` name outside of the function. When you declare a variable (like here) inside a function, it's name will override for the scope of the function, the name of any variables declared outside.

This means that although it looks like `resFromAPI` is the same thing on line 6 and 10, that's not guaranteed. They are only the same thing if you pass `resFromAPI` from the outer scope to the inner scope as a parameter on line 20.

Imagine changing line 20 to `sortRatingArray([])`. Now you're passing an empty array, which means that the `resFromAPI´ variable inside the function will be an empty array, while the `resFromAPI` variable outside of the function is something else.

To summarize: Avoid reusing the same name in two different scopes that are accessible to each other. It's legal, but can cause confusion, and often it doesn't make the most sense anyway since the two scopes probably have different perspective on the variable anyway.

**Review by richardolsson — COMMENTED** [review]
Nice way of refactoring the code to be able to reuse the `showAllChallenges()` function. I've added a bunch of comments below. Let me know if you have any questions!
